packages/exhibitor/extra/bootstrap_exhibitor_tls.py

```diff
@@ -0,0 +1,213 @@
+        ],
+        stderr=subprocess.STDOUT,
+    )
+    print(output.decode())
+
+    print('Making CSR for {} with IP `{}`'.format(client_entity, ip))
+    output = subprocess.check_output(
+        args=[
+            BOOTSTRAP_CA_BINARY, 'csr', client_entity,
+            '--output-dir', EXHIBITOR_TLS_TMP_DIR,
+            '--url', ca_url,
+            '--ca', CSR_SERVICE_CERT_PATH,
+            '--psk', psk,
+            '--sans', '{},localhost,127.0.0.1,exhibitor'.format(ip),
+        ],
+        stderr=subprocess.STDOUT,
+    )
+    print(output.decode())
+
+    print('Writing TLS artifacts to {}'.format(artifacts_path))
+    output = subprocess.check_output(
+        args=[
+            BOOTSTRAP_CA_BINARY, 'create-exhibitor-artifacts',
+            '--output-dir', EXHIBITOR_TLS_TMP_DIR,
+            '--ca', CSR_SERVICE_CERT_PATH,
+            '--client-entity', client_entity,
+            '--server-entity', server_entity,
+            '--artifacts-directory', '{}'.format(artifacts_path),
+        ],
+        stderr=subprocess.STDOUT,
+    )
+    print(output.decode())
+
+
+def _fail_and_calculate_status(message):
+    exit_status = 1 if os.environ.get('EXHIBITOR_TLS_REQUIRED') == 'true' else 0
+    if exit_status:
+        print('exhibitor TLS bootstrap failed, but is required')
+    else:
+        print('exhibitor TLS bootstrap failed, launching in insecure mode')
+    print(message)
+    sys.exit(exit_status)
+
+
+def _remove_artifact_dir():
+    p = Path(TLS_ARTIFACT_LOCATION)
+    if p.is_dir():
+        for f in p.iterdir():
+            f.unlink()
+        p.rmdir()
+
+
+def main():
+    if os.environ.get('EXHIBITOR_TLS_ENABLED', 'false') == 'false':
+        _fail_and_calculate_status('exhibitor TLS is disabled')
+
+    if Path(TLS_ARTIFACT_LOCATION).exists():
+        return
+
+    if not Path(CSR_SERVICE_CERT_PATH).exists():
+        _fail_and_calculate_status('root CA certificate does not exist')
+
+    exhibitor_bootstrap_ca_url = os.environ['EXHIBITOR_BOOTSTRAP_CA_URL']
+    bootstrap_url = os.environ['BOOTSTRAP_URL']
+
+    ca_url = get_ca_url(exhibitor_bootstrap_ca_url, bootstrap_url)
+    if not ca_url:
+        _fail_and_calculate_status('could not determine ca service url')
+
+    if not test_connection(ca_url):
+        _fail_and_calculate_status('connection failed')
+
+    print('Bootstrapping exhibitor TLS')
+
+    try:
+        gen_tls_artifacts(ca_url, Path(TLS_ARTIFACT_LOCATION))
+    except subprocess.CalledProcessError as cpe:
+        # Clean up any partially written artifacts
+        _remove_artifact_dir()
+        _fail_and_calculate_status(
+            'error generating artifacts code: {} stdout: {} stderr: {}'.format(
+                cpe.returncode, cpe.stdout, cpe.stderr))
+    # remove file from temporary location
+    Path(CSR_SERVICE_CERT_PATH).unlink()
+
+
+if __name__ == '__main__':
+    main()
+    # Always flush stdout buffer when exiting the script
+    sys.stdout.flush()
```

Comment:
mhm if this has been a problem I recommend starting the Python interperter with the environment variable PYTHONIOBUFFERING set so that std stream buffering gets disabled completely.